cmake/embed_helpers.cmake

```diff
@@ -0,0 +1,118 @@
+
+function(prepare_clang_patches patch_command)
+  message("Building embedded Clang against host LLVM, checking compatibiilty...")
+  detect_os()
+  message("HOST ID ${HOST_OS_ID}")
+
+  set(CLANG_PATCH_COMMAND "/bin/true")
+  if(HOST_OS_ID STREQUAL "debian" OR HOST_OS_ID STREQUAL "ubuntu" OR HOST_OS_ID_LIKE STREQUAL "debian")
+    message("Building on a debian-like system, will apply minimal debian patches to clang sources in order to build.")
+    set(PATCH_NAME "debian-patches.tar.gz")
+    set(PATCH_PATH "${CMAKE_CURRENT_BINARY_DIR}/debian-llvm/")
+    set(DEBIAN_PATCH_SERIES "")
+    list(APPEND DEBIAN_PATCH_SERIES "kfreebsd/clang_lib_Basic_Targets.diff -p2")
+
+    if(${LLVM_VERSION} VERSION_EQUAL "8" OR ${LLVM_VERSION} VERSION_GREATER "8" )
+      set(DEBIAN_PATCH_URL_BASE "https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/archive/debian/")
+      set(DEBIAN_PATCH_URL_PATH "8_8.0.1-1/llvm-toolchain-debian-8_8.0.1-1.tar.gz?path=debian%2Fpatches")
+      set(DEBIAN_PATCH_URL "${DEBIAN_PATCH_URL_BASE}/${DEBIAN_PATCH_URL_PATH}")
+      set(DEBIAN_PATCH_CHECKSUM 2b845a5de3cc2d49924b632d3e7a2fca53c55151e586528750ace2cb2aae23db)
+    else()
+      message(FATAL_ERROR "No supported LLVM version has been specified with LLVM_VERSION (LLVM_VERSION=${LLVM_VERSION}), aborting")
+    endif()
+
+    list(LENGTH DEBIAN_PATCH_SERIES NUM_PATCHES)
+    message("${NUM_PATCHES} patches will be applied for Clang ${LLVM_VERSION} on ${HOST_OS_ID}/${HOST_OS_ID_LIKE}")
+    fetch_patches(${PATCH_NAME} ${PATCH_PATH} ${DEBIAN_PATCH_URL} ${DEBIAN_PATCH_CHECKSUM})
+    prepare_patch_series("${DEBIAN_PATCH_SERIES}" ${PATCH_PATH})
+
+    # These targets are from LLVMExports.cmake, so may vary by distribution.
+    # in order to avoid fighting with what the LLVM package wants the linker to
+    # do, it is easiest to just override the target link properties
+
+    # These libraries are missing from the linker line command line
+    # in the upstream package
+    # Adding extra libraries here shouldn't affect the result, as they will be
+    # ignored by the linker if not needed
+    set_target_properties(LLVMSupport PROPERTIES
+      INTERFACE_LINK_LIBRARIES "LLVMCoroutines;LLVMCoverage;LLVMDebugInfoDWARF;LLVMDebugInfoPDB;LLVMDemangle;LLVMDlltoolDriver;LLVMFuzzMutate;LLVMInterpreter;LLVMLibDriver;LLVMLineEditor;LLVMLTO;LLVMMCA;LLVMMIRParser;LLVMObjCARCOpts;LLVMObjectYAML;LLVMOption;LLVMOptRemarks;LLVMPasses;LLVMPerfJITEvents;LLVMSymbolize;LLVMTableGen;LLVMTextAPI;LLVMWindowsManifest;LLVMXRay;-Wl,-Bstatic -ltinfo;"
+    )
+
+    # Need to omit lpthread here or it will try and link statically, and fail
+    set_target_properties(LLVMCodeGen PROPERTIES
+      INTERFACE_LINK_LIBRARIES "LLVMAnalysis;LLVMBitReader;LLVMBitWriter;LLVMCore;LLVMMC;LLVMProfileData;LLVMScalarOpts;LLVMSupport;LLVMTarget;LLVMTransformUtils"
+    )
+
+    set(CLANG_PATCH_COMMAND "(QUILT_PATCHES=${PATCH_PATH} quilt push -a || [[ $? -eq 2 ]])")
+  endif()
```

Comment:
I'm ok with this for now (since it's pretty modular and hidden behind a flag) but I am concerned this will be painful in the long term. For example, what happens if debian's llvm8 patches don't apply cleanly to llvm9? And what happens when llvm10 is out? I worry that for every llvm version we support we'll have an `#patches * #support_llvm_versions` custom patches to keep track of.

Reply:
> llvm8 patches don't apply cleanly to llvm9

Luckily this is actually pulling patches directly from LLVM's own build repo, which is linked to by https://apt.llvm.org/. You can see the branch associated with the LLVM version.

This shows the patch series used to build LLVM 9 for debian. In the case of LLVM 8, to get libclang.a to build I only needed to apply 1 single patch, not the whole series. The worst possible case is you need to copy the series file, and set "-p2" or "-p1" depending on if it is LLVM itself or a subproject, and what patch level it was written at.

They are patching the same vanilla sources and both use quilt, so I figured it shouldn't be too bad. The workflow on a new LLVM release would be to:

1. Try it with the existing patch series and maybe it just works :tada: 
1. Examine the build failure, and grep for patches affecting these files from those in the patch series on the LLVM 9 / 10 etc branch
1. Add patches to the series file until there is a minimal one that builds.

Most of the time the patches change minor cosmetic things, no major functionality. In this case, the patch makes it from `KFreebsd` to `kFreebsd` which breaks a header, but it is easy to fix. Hopefully the future isn't too bad. I definitely share your concerns though, that's why I tried to make this as simple and module as possible.

I expect other sources will need some patches applied to them too, eg libelf uses composite functions which aren't supported on clang. I saw a patch to chromiumos that just moved the composite functions to static. Writing this patch would make the process of building libelf much simpler, and leverage this existing simple patching faculty using quilt.

Reply:
Ah didn't realize llvm team was supporting the patches. In that case it should be fine.

Reply:
> Ah didn't realize llvm team was supporting the patches. In that case it should be fine.

I'll document the patches as thoroughly as I can, and point out quirks. I'll consolidate this in the embedded_builds doc.

Reply:
Added this to the docs